python/ray/node.py

```diff
@@ -144,6 +148,16 @@ def __init__(self,
             ray_params.include_java = (
                 ray.services.include_java_from_redis(redis_client))
 
+        if head or not connect_only:
+            # We need to start a local raylet.
+            if (self._ray_params.node_manager_port is None
+                    or self._ray_params.node_manager_port == 0):
+                # No port specified. Pick a random port for the raylet to use.
+                # NOTE: There is a possible but unlikely race condition where
+                # the port is bound by another process between now and when the
+                # raylet starts.
+                self._ray_params.node_manager_port = self._get_unused_port()
```

Comment:
Currently worker talks to raylet via UNIX domain socket, and grpc supports that too, that UNIX socket name should not conflict with other applications. Do we want to use TCP socket for local worker connection also? (which is fine, but just want to confirm). 

And if the intent is to let the driver know the port of the raylet, maybe we could instead let driver do a sync read to get the raylet port from GCS, as it needs to connect to GCS anyway.

Reply:
Yes, we're using TCP just to make all worker-raylet connections uniform. Eventually, we would like to submit tasks directly from a worker to a remote raylet too.

Hmm I thought about doing it with sync reads to the GCS, but decided against it. For testing purposes, we often connect multiple raylets to the GCS on a single node, so we'd have to differentiate between them to find the right raylet. It's a bit of a headache.

Reply:
@stephanie-wang How about we choose a random node ID instead and pass that to the driver? then driver can wait for this node ID to appear on GCS and get the port for the specific raylet then. There would be no collisions in this case.

Reply:
I might do that for #6143, but I'll hold off on getting the raylet port from the GCS. I don't think it's a big deal to set the port from the driver for now.